### PR TITLE
[library] Allow require to load libraries from memory.

### PR DESCRIPTION
**Pull request by ejgallego**
This is one more step in our quest to avoid saving / loading .vo files to disk when not needed.

Most changes are straightforward except for the extension of the command type to include an internalization function.

This is IMHO a good start to the goal of having the vernacular layer access all file-system by a parameterized structure to interpretation, so document managers can properly handle file-system access.

For example, a typical use case in jsCoq is to download a .vo file on Require, but there are many more possibilities.

Overlays:

- https://github.com/coq-community/vscoq/pull/775
- https://github.com/ejgallego/coq-lsp/pull/738

### CODE REVIEW

**Comment by ejgallego** [comment]
@coqbot run full CI

File: vernac/library.ml
Comment (SkySkimmer) [inline vernac/library.ml:290]: I would move the `if not (DirPath.equal dir lsd.md_name)` (now `.library_name`) here since `intern` is not necessarily `intern_from_file`

File: vernac/library.ml
Comment (SkySkimmer) [inline vernac/library.ml:260]: Considering how `lib_resolver` and `dir` are used I think it would make sense to have `intern_file : string -> library_t`
Then some upper layer would define 
~~~ocaml
let intern_from_file dir =
  let f = Loadpath.try_locate_absolute_library dir in
  Feedback.feedback(Feedback.FileDependency (Some f, DirPath.to_string dir));
  let lib = Library.intern_file f in
  Feedback.feedback (Feedback.FileLoaded(DirPath.to_string dir, f));
  lib
~~~
(`DirPath.equal dir lsd.md_name` check left to `intern_library` as per my other comment)
(this would be done somewhere where we don't need to abstract over lib_resolver, I guess sysinit since stm depends on sysinit and sysinit needs this API)

**Comment by SkySkimmer** [comment]
@ejgallego ping

**Comment by gares** [comment]
The entire patch makes it possible to change the `~intern` value at run time.

I don't see a use case, and we fix it once and for all at system initialization time, then this patch becomes much simpler, e.g. no need to changes to vernacextend.

Is there any use case for not having the intern function be given only once?

**Review by gares — CHANGES_REQUESTED** [review]
simplify the patch making the intern argument part of the system configuration

**Comment by ejgallego** [comment]
> The entire patch makes it possible to change the ~intern value at run time.
>
> I don't see a use case, and we fix it once and for all at system initialization time, then this patch becomes much simpler, e.g. no need to changes to vernacextend.
>
> Is there any use case for not having the intern function be given only once?

The main use case here is not so much to dynamically update the .vo intern function, but to follow a methodology where effects are reflected in the types of the commands. In this case, I'd like to have all the commands that require access to the env FS to reflect so in the type so the incremental engine can properly track these dependencies without having to analyze the internals.

File: vernac/library.ml
Comment (ejgallego) [reply vernac/library.ml:306]: Indeed, I've done as you propose. However see my comment.

Comment (ejgallego) [reply vernac/library.ml:306]: I cannot post a comment where I wanted, but indeed see this code:
```ocaml
    user_err
      (str "The file " ++ str file ++ str " contains library" ++ spc () ++
       DirPath.print library_name ++ spc () ++ str "and not library" ++
       spc() ++ DirPath.print dir ++ str ".")
```
Indeed, it makes sense to move it, however we will need a way to find the provenance from the parametric `intern` function.

Comment (ejgallego) [reply vernac/library.ml:306]: I assume you'd be OK to do like that, right?

File: vernac/library.ml
Comment (ejgallego) [reply vernac/library.ml:273]: It makes sense; we could just get rid of the `Feedback.File` stuff altogether (it was nice for jsCoq), as now users can just provide their own `intern` function, which should be usable for this.

This PR is usable for my purposes, however the `Loadpath` API is not.

**Comment by ejgallego** [comment]
> simplify the patch making the intern argument part of the system configuration

Review is not relevant anymore, parsing and separation PR moved the intern phase to synterp and we don't have classifier for synterp vernacs in terms of the effects they produce.

I've workarounded this in my implementation by using some minimal Ast analysis (far from ideal, but should work OK at least for this case)

**Comment by gares** [comment]
AFAICT you are still passing a parameter around. The review asks to pass the parameter once and for all to system initialization. I believe it is still relevant.

**Comment by SkySkimmer** [comment]
@gares so add a global `Intern.t ref` to Library instead of passing it as a parameter? That is also OK for me.

**Comment by gares** [comment]
Which is the use case for changing this parameter at run time?

**Comment by ejgallego** [comment]
> Which is the use case for changing this parameter at run time?

I'm not sure yet, I am experimenting with some different approaches.

**Comment by SkySkimmer** [comment]
@gares at the coq call we could not figure out the need for your request for changes.
Please explain more, or alternatively we can merge as is (after rebase of course).

**Comment by gares** [comment]
It Is a matter of convenience and self documentation. Something set up at coq initialization time is supposed not to change, something you explicitly pass around may change but you pay the price of verbosity.

My last message was asking for use cases where the loader would change at run time. We're these provided in the call?

**Comment by ejgallego** [comment]
> It Is a matter of convenience and self documentation. 

How is it setting a global reference more convenient or better documented that declaring the effects a function perform in the type?

> Something set up at coq initialization time is supposed not to change, something you explicitly pass around may change but you pay the price of verbosity.
>
> My last message was asking for use cases where the loader would change at run time. We're these provided in the call?

There are use cases where the loader could change, but that's not the point of the current design, as I already explained above a few times.

Reading a .vo file is a side effect that needs to be tracked by the upper layers. It is good methodology to force commands accessing the filesystem to declare this access, so random commands cannot start reading .vo files for example without the upper layers being aware of it.

In this case the "price of verbosity" is pretty small, as compared to the gain of having a more principled API.

**Comment by gares** [comment]
So there are no concrete use cases you can cite.

**Comment by ejgallego** [comment]
> So there are no concrete use cases you can cite.

My main use case is detecting which vernacular functions do read .vo files.

**Comment by gares** [comment]
This is not a *use* case.

**Comment by ejgallego** [comment]
> This is not a _use_ case.

Ok, if you don't want to call it "use case" is fine to me; but this is still an important need.

**Comment by SkySkimmer** [comment]
This argument about changing the value at runtime is nonsense, both a global ref and a passed parameter can do it.

**Comment by gares** [comment]
A global ref you set once at initialization time (the sysinit component) is not really a thing meant to be changed at runtime, even if ocaml let's you do it.

My impression from this discussion is that there is no need to pass the parameter, since it does not change while coq runs. If there was a usecase, I would prefer the explicit parameter.

Does it still sound like nonsense?

**Comment by ejgallego** [comment]
> Does it still sound like nonsense?

You seem to focus on some criteria about refs vs parameter for things that change vs not. That's not the point of this PR.

The point of the technical solution chosen is to track which commands are going to access .vo reading, so that if a command needs to read a .vo, it has to declare it in the type.

This is an important need.

**Comment by SkySkimmer** [comment]
>My impression from this discussion is that there is no need to pass the parameter, since it does not change while coq runs. If there was a usecase, I would prefer the explicit parameter.

I don't really understand the preference TBH.

Also I feel like this discussion should have been handled as "make suggestion for change -> Emilio rejects it because he really wants the parameter -> have it your way" instead of blocking the PR.

**Review by gares — APPROVED** [review]
We can move on.

File: vernac/library.ml
Comment (ejgallego) [inline vernac/library.ml:540]: @ppedrot can you confirm this is correct w.r.t. the new `Vmlibrary` stuff?

Comment (ppedrot) [reply vernac/library.ml:540]: I think so. Unrelatedly the dummy digest part is slightly more concerning to me, we should have an option there or something to signal that the data didn't come from a vo file.

Comment (ppedrot) [reply vernac/library.ml:540]: (Also as-is this will never work with native, which is another reason why we should have a distinct library type.)

Comment (ejgallego) [reply vernac/library.ml:540]: Let me indeed fix the digest stuff before this is merged.

Actually native could be made work with this in the same way: have the kernel relay a .cmxs request to the document manager / whoever is coordinating the filesystem layout.

Indeed it seems to me that a few parts of Coq could benefit from exposing a caching layer (that's the way I see the .vo files, as a cache), for example in #17674 ; I wanted to discuss this in the call actually.

**Comment by ppedrot** [comment]
Can we try to get this merged quickly before another flamewar roars up?

**Comment by ejgallego** [comment]
@ppedrot unfortunately it's gonna take me a while to rebase and test given my current workload; feel free to take over with the rebase + fixes due to upstream changes; I can do the testing myself.

**Comment by ppedrot** [comment]
I've performed the rebase. @coqbot run full ci

File: vernac/library.ml
Comment (SkySkimmer) [inline vernac/library.ml:504]: Is removing the dummy digest still todo?

Comment (ejgallego) [reply vernac/library.ml:504]: @ppedrot what should we do here?

Comment (ppedrot) [reply vernac/library.ml:504]: I'd put an option type and call it a day, or even just add a constructor to Safe_typing.vodigest.

Comment (ejgallego) [reply vernac/library.ml:504]: Ok, let me understand what purpose this thing serves, and I'll update it.

Comment (ppedrot) [reply vernac/library.ml:504]: This is used to check that the vo files didn't get modified under our feet. If you want added safety you could try to generate a unique number in the safe_typing API and check that instead of the hash of the data.

**Comment by ejgallego** [comment]
Ok, overlays done. I will fix the digest nit while the full CI runs.

@coqbot run full CI

File: vernac/library.ml
Comment (ejgallego) [reply vernac/library.ml]: I've opted to add a digest of the data, need to do some testing tho.

**Comment by ejgallego** [comment]
@coqbot run full CI

**Comment by ejgallego** [comment]
I think this is OK now.

**Comment by ejgallego** [comment]
Gentle ping @SkySkimmer

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**
